npm/webpack-dev-server/src/IndexHtmlWebpackPlugin.ts

```diff
@@ -0,0 +1,81 @@
+import type { Compiler, Compilation } from 'webpack'
+import * as fs from 'fs/promises'
+import * as path from 'path'
+import type { CreateFinalWebpackConfig } from './createWebpackDevServer'
+
+const extensionRegexp = /\.(css|js|mjs)(\?|$)/
+
+export class CypressIndexHtmlWebpackPlugin {
+  private webpack = this.config.sourceWebpackModulesResult.webpack.module as any
+
+  constructor (public config: CreateFinalWebpackConfig, public publicPath: string) {}
+
+  apply (compiler: Compiler): void {
+    if (this.config.sourceWebpackModulesResult.webpack.majorVersion === 4) {
+      compiler.hooks.emit.tapAsync('CypressIndexHtmlWebpackPlugin', async (compilation, callback) => {
+        const htmlContent = await this.transformEntryPoints(compilation, compiler)
+
+        const assets = compilation.assets as any
+
+        assets['index.html'] = {
+          source: () => htmlContent,
+          size: () => htmlContent.length,
+        }
+
+        callback()
+      })
+    } else {
+      compiler.hooks.thisCompilation.tap('CypressIndexHtmlWebpackPlugin', (compilation) => {
+        compilation.hooks.processAssets.tapPromise({ name: 'CypressIndexHtmlWebpackPlugin' }, async (assets) => {
+          const htmlContent = await this.transformEntryPoints(compilation, compiler)
+          const htmlSource = new this.webpack.sources.RawSource(htmlContent)
+
+          assets['index.html'] = htmlSource
+        })
+      })
+    }
+  }
+
+  private async transformEntryPoints (compilation: Compilation, compiler: Compiler): Promise<string> {
+    const entryNames = Array.from(compilation.entrypoints.keys())
+    const entryPointPublicPaths: string[] = []
+
+    for (const entryName of entryNames) {
+      const entryPointFiles = compilation.entrypoints.get(entryName)?.getFiles() || []
+      const publicPaths = entryPointFiles
+      .map((chunkFile) => this.publicPath + urlEncodePath(chunkFile))
+      .filter((file) => extensionRegexp.test(file))
+
+      entryPointPublicPaths.push(...publicPaths)
+    }
+
+    const indexHtmlContent = await this.getIndexContent(compilation, compiler)
+    const entryPointScriptTags = entryPointPublicPaths.map((entryPath) => `<script defer src="${entryPath}"></script>`).join('')
+    const indexContentWithScripts = indexHtmlContent.replace(/<\/head>/, `${entryPointScriptTags}</head>`)
+
+    return indexContentWithScripts
+  }
+
+  private async getIndexContent (compilation: Compilation, compiler: Compiler): Promise<string> {
```

Comment:
Cool, nice to see you working on this. html webpack plugin does a lot of things - do we know what things it does that we don't need (eg, what makes it overkill?)

Reply:
HtmlWebpackPlugin contains a whole bunch of [options](https://github.com/jantimon/html-webpack-plugin#options) that we don't need and don't surface to the user. The only thing we rely on is the live-reloading and the injection of script tags. The one thing that we would lose is the handlebar templating syntax that HtmlWebpackPlugin handles.

I could change the flow a bit and have this plugin be a fallback in case we can't `require('html-webpack-plugin')`. If the user doesn't have this, they would just need a barebones plugin which we would provide.